keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -1188,239 +937,309 @@ drivers are expected to be written for Kubernetes.
 
 ```go
-// ResourceSlice provides information about available
-// resources on individual nodes.
+// One of more slices represent a pool of devices managed by a given driver. For
+// node-local devices, each node is a pool. How many slices the driver uses
+// to publish that pool is driver-specific. Within each pool the device name must
+// be unique.
+//
+// In which slice a device gets published may change over time. The unique identifier
+// for a node-local device is the tuple `<driver name>/<node name>/<device name>`. Each
+// of these names is a DNS label or domain, so it is okay to concatenate them
+// like this in a string with a slash as separator.
+//
+// Whenever a driver needs to update a pool, it bumps the pool generation number
+// and updates the slices with that new number and new device definitions. A consumer
+// must only use device definitions from the slices with the highest generation number
+// and ignore the others.
+//
+// If necessary, a consumer can check the number of total devices in a pool (included
+// in each slice) to determine whether its view of a pool is complete.
+//
+// For devices that are not local to a node, the node name is not set. Instead,
+// a pool name is chosen by the driver if necessary. It can be left blank if the
+// device names are unique in the entire cluster. The unique identifier then
+// has the same format as before with `<driver name>[/<pool name>]/<device name>`.
 type ResourceSlice struct {
     metav1.TypeMeta
     // Standard object metadata
+    // +optional
     metav1.ObjectMeta
 
-    // NodeName identifies the node which provides the resources
-    // if they are local to a node.
+    Spec ResourceSliceSpec
+
+    // A status might get added later.
+}
+
+type ResourceSliceSpec struct {
+    // NodeName identifies the node which provides the devices. All devices
+    // are local to that node and part of a per-node pool.
     //
-    // A field selector can be used to list only ResourceSlice
-    // objects with a certain node name.
+    // This field can be used to limit access from nodes to slices with
+    // the same node name. It also indicates to autoscalers that adding
+    // new nodes of the same type as some old node might also make new
+    // devices available.
+    //
+    // NodeName and PoolName are mutually exclusive.
+    //
+    // +optional
     NodeName string
 
+    // PoolName can be used to describe pools of devices which are not local
+    // to a node. It can be left blank if the device names are unique in the cluster.
+    //
+    // It must not be longer than 253 and must consist of one or more DNS sub-domains
+    // separated by slashes.
+    //
+    // +optional
+    PoolName string
+
+    // The generation gets bumped in all slices of a pool whenever device
+    // definitions changed. A consumer must only use device definitions from the slices
+    // with the highest generation number and ignore the others.
+    PoolGeneration int64
```

Comment:
TODO (can be decided during implementation): can `metadata.generation` be used instead of creating a new field?

I tried with the current `ResourceSlice` implementation in 1.30 and could set `metadata.generation` on create, but not bump it during patching (tried `kubectl apply` and `kubectl apply --server-side`). Perhaps it would work if the type had a different support for the field.

Reply:
No, `metadata.generation` on updates is not possible:
https://github.com/kubernetes/kubernetes/blob/6ba9fa89fb5889550649bfde847c742a55d3d29c/staging/src/k8s.io/apiserver/pkg/registry/rest/update.go#L122-L127

That seems mandatory for all resources, so we have to use a separate field for this cross-resources generation.